In [1]:
import numpy as np
import math as mt

<h2>Sigmoid function</h2>

In [44]:
# x=Dot product of X and theta
def sigmoid(x):
    return 1/(1+np.exp(-x))

In [ ]:
# Here Theta1 and Theta2 are pre-trained. 

In [3]:
import scipy.io

In [4]:
mat=scipy.io.loadmat('ex3data1.mat')

In [5]:
X=np.array(mat['X'])

In [6]:
y=np.array(mat['y'])

In [7]:
X.shape

(5000, 400)

In [8]:
#adding column of 1's
X=np.hstack((np.ones((X.shape[0],1)),X))

In [9]:
X.shape

(5000, 401)

In [10]:
y.shape

(5000, 1)

In [11]:
mat_weights=scipy.io.loadmat('ex3weights.mat')

In [12]:
Theta1=mat_weights['Theta1']
Theta2=mat_weights['Theta2']

In [13]:
Theta1.shape

(25, 401)

In [14]:
Theta2.shape

(10, 26)

<h2>Predicting function</h2>

In [ ]:
#Suppose that we have input layer with 400 units, one hidden layer with 25 units and final output layer with 10 output units
#Theta1 and Theta2 are already trained. We will just predict here.
#This code is for classification of handwritten digits which has 5000 records each having 4000 features of a digit.

In [29]:
def predict(Theta1,Theta2,X):
    a=sigmoid(np.dot(X,np.transpose(Theta1)))
    a=np.hstack((np.ones((a.shape[0],1)),a))
    prediction=sigmoid(np.dot(a,np.transpose(Theta2)))
    predictions=prediction.argmax(axis=1)
    return predictions+1
    
predictions=predict(Theta1,Theta2,X)
    
    
        

In [31]:
from sklearn import metrics

In [32]:
metrics.accuracy_score(y,predictions) #checking accuracy score

0.9752

<h2>Yayyyy!! We got 97.52% of accuracy in Recognizing Handwritten digits</h2>

<h3>Sigmoid gradient</h3>  

In [50]:
def sigmoidGradient(z):
    g=sigmoid(z)*(1-sigmoid(z))
    return g
    
sigmoidGradient(0)   

0.25


<h3>Initializing weights</h3>  

In [67]:
# This will return a numpy array with thetas between [-E,E]
# E =epsilon
epsilon_init=0.12
def RandomInitializeWeights(L_out,L_in):
    W=np.random.uniform(0,1,(L_out,L_in+1))*2*epsilon_init-epsilon_init
    return W
    

In [68]:
RandomInitializeWeights(4,5)

array([[-0.06133985,  0.08655642,  0.02278681, -0.06166842, -0.05541482,
         0.02966625],
       [-0.02460955, -0.00727979, -0.04849956, -0.10678992, -0.02136458,
        -0.06312647],
       [-0.09783017,  0.11548449, -0.09752824,  0.00271352, -0.00195583,
         0.10102853],
       [-0.00971861, -0.08492481, -0.11175609, -0.07593392,  0.05402399,
        -0.02578409]])

<h2>Neural Network Cost Function</h2>

In [76]:
Theta1.shape

(25, 401)

In [77]:
Theta2.shape

(10, 26)

In [78]:
X.shape

(5000, 401)

In [79]:
y.shape

(5000, 1)

<h3>Forward Propogation</h3>

In [92]:
def predict(X,Theta1,Theta2):
    a=sigmoid(np.dot(X,np.transpose(Theta1)))
    a=np.hstack((np.ones((a.shape[0],1)),a))
    predict=sigmoid(np.dot(a,np.transpose(Theta2)))
    return predict
    

In [93]:
y_pred=predict(X,Theta1,Theta2)

In [94]:
y_pred.shape

(5000, 10)

In [95]:
y.shape

(5000, 1)

In [ ]:
# Before finding the cost function we have to convert our y to the shape of y_pred 

In [96]:
y_converted=np.zeros(y_pred.shape)

In [100]:
for i in range(len(y)):
    y_converted[i,y[i]-1]=1
    

In [111]:
# Now let us find the cost function
m=len(X)

In [113]:
J=sum(sum(y_converted*np.log(y_pred)+(1-y_converted)*np.log(1-y_pred)))*(-1/m)

In [114]:
J

0.28762916516131876

In [116]:
a=sigmoid(np.dot(X,np.transpose(Theta1)))
a=np.hstack((np.ones((a.shape[0],1)),a))
predict=sigmoid(np.dot(a,np.transpose(Theta2)))

<h3>Backward Propogation</h3>

In [118]:
error3=y_pred-y_converted
error2=np.dot(error3,Theta2)*(a)*(1-a)
error2=error2[:,1:]
Theta1_grad=np.zeros(Theta1.shape)
Theta2_grad=np.zeros(Theta2.shape)

In [122]:
# Thats the gradient of cost function

In [123]:
Theta1_grad=(Theta1_grad+np.dot(np.transpose(error2),X))*(1/m)
Theta2_grad=(Theta2_grad+np.dot(np.transpose(error3),a))*(1/m)

(5000, 25)

In [ ]:
# Further perform gradient descent in the similar vway we did in logistic regression. 
# If you don't know how to do then check my logistic regression with python repository